[2단계 - 로또 리팩토링] 가콩(최가빈) 미션 제출합니다.

gabean13 opened the pull request:
안녕하세요 지노😀 
2단계 리뷰 부탁드립니다

### step 1 코멘트
---
❓Dto는 어떤 경우에 사용하시나요? 의존성을 보니 도메인 모델이 의존하는 경우도 있고 도메인 모델의 dto가 의존하고도 있어서요.
- Dto는 도메인의 로직이 필요없이 데이터를 전달할때만 사용하려고 했습니다. (view).

❓출력을 위해서는 getter를 사용하시면 될 것 같아요. 현재는 콘솔 환경이지만 웹 환경이라면 toString은 유효할까요?
- 넵 보고 toString은 삭제하고 dto를 만들어서 관리했습니다. toString은 디버깅에서 사용하고, 출력에서는 dto, getter를 사용하도록 하겠습니다!

❓ 중복을 허용하지 않는 자료구조는 없을까요?
- Set으로 수정했습니다!

❓. 로또 생성이 랜덤에 의존하고 있는데요. 랜덤은 테스트하기 어렵습니다. 어떻게 하면 테스트하기 쉽게 변경할 수 있을까요?
- interface를 만들고, 실제 코드에서는 랜덤으로 로또를 생성하는 구현체를 사용하고, 테스트에서는 고정된 값을 반환할 수 있는 구현체를 사용하여 테스트하였습니다. 
- interface : RandomNumberGenerator
- production : LottoNumberGenerator
- test : FixedNumberGenerator

❓12는 어떤 의미일까요?
- 두 Set을 합친 뒤, 2개의 중복되는 값들은 1개의 값으로 통일 될 것이라 겹치는 개수를 세기 위한 숫자였습니다. 만약 하나도 겹치지 않는다면 12, 전부 겹친다면 6입니다. 의미를 좀 더 이해하기 쉽게 MAX_UNIQUE_NUMBERS로 상수화 하였습니다

❓개인적으로는 좋은 방법은 아닌 것 같아요. 필드에 선언된 ranks 하나면 수익률을 구할 수 있지 않나요?
- 네! ranks만 필드로 가지도록 수정하고 profitRate는 출력할때만 반환하도록 변경하였습니다

❓해당 클래스의 의도는 무엇일까요? purchaseNumber를 필드에 가지고 있어야 하나요?
- UserLotto 클래스의 의미는 user가 생성한 lotto들을 가지고 있고, 해당 lotto들로 결과를 계산하는 의도를 가지도록 했습니다.
- purchaseNumber는 lottos의 size에 로또 가격을 곱하면 바로 구할 수 있어서 필요할 때 사용하도록 메서드로 분리하고 필드는 삭제했습니다!

❓WinningLotto의 당첨 번호 6개는 Lotto 라고 볼 수 있지 않을까요?
- 네! 그래서 WinningLotto의 winninNumbers를 Lotto로 선언하였습니다. 그리고 기존에는 winningNumbers에서 입력된 로또들의 형식을 검사하였는데 lotto의 성질과 동일하여 lotto에서 해당 검사를 하도록 변경하였습니다.

❓WinningLotto의 생성과 별개로 bonus를 채워주는 이유가 따로 있을까요? 생성과 동시에 모든 필드에 할당 될 수는 없을까요?
- 원래 동시에 bonus를 채울려했는데요, 그렇게 하면 보너스 번호가 틀리면 당첨번호를 다시 입력받아야하는 불편함이 있을 것 같아서 별개로 채워주었습니다. 그런데 이번에 리팩토링하면서 bonus를 final로 선언하여 값이 변하지 않도록 하는 것이 당첨번호를 한번 더 입력하는 것보다 더 
중요하다고 판단되어서 생성하는 동시에 채워주도록 변경하였습니다.

❓개인적으로 inputView에서 데이터 타입 정도는 변환해도 괜찮다고 생각해요.
- 네! 생각해보니 타입 변환을 굳이 여러 번 할 필요 없다고 판단돼서 inputValidator에서 유효성 검사를 통과하면 데이터 타입을 변환하여 반환하도록 수정하였습니다.

리뷰 감사합니다
좋은 하루 되세요 👍🏻

gabean13 committed:
refactor(Lotto) : lotto number random 생성 로직 변경

- RadomNumberGenerator를 받아서 LottoNumberGenerator에서 Lotto 번호 생성
- LottoController, LottoApplication, UserLotto에 RandomNumberGenerator 의존 추가

gabean13 committed:
test(LottoNumberGenerator) : 로또 랜덤 생성 방식 변경에 따른 테스트 변경 및 테스트 추가

gabean13 committed:
feat(Input) : input view에서 데이터 타입 변환하여 반환

- purchaseAmount, winningNumber, bonus가 형식이 맞는 경우 데이터 타입 변환
- bonus 형식 검사 추가

gabean13 committed:
refactor(OutputView) : dto를 이용하여 출력하도록 수정, toString 사용 삭제

gabean13 committed:
refactor(WinningLotto) : 생성할 때 bonus, winningNumber 할당하도록 변경

- winningNumber Lotto 객체로 변경
- 보너스 번호 일치, lotto 번호 비교하여 중복되는 숫자 찾는 메서드 추가
- lotto class에서 객체 생성할 때 확인하는 유효성 검사 메서드들은 삭제

gabean13 committed:
feat(UserLotto) : userLotto에서 당첨 통계 계산

gabean13 committed:
refactor(LottoResult) : profitRate 변수 삭제 및 LottoResultDto 추가

gabean13 committed:
feat(Lotto) : lotto 객체 내에서 로또 형식 유효성 검사

gabean13 committed:
test(LottoNumberGenerator) : 무작위 정수 생성 테스트

gabean13 committed:
test(LottoResult) : 로또 당첨 결과 테스트

gabean13 committed:
test(Lotto) : lotto 형식 검사 테스트

gabean13 committed:
test(UserLotto) : 1등 ~ 낙첨까지 결과 테스트

gabean13 committed:
test(WinningLotto) : 보너스 볼 형식 검사, 보너스 볼 유무 검사 테스트

gabean13 committed:
test(InputValidator) : 보너스 번호, 당첨 번호 숫자 여부 검사 추가

gabean13 committed:
docs(README) : 변경된 기능 요구사항 업데이트

hyunssooo commented on src/main/java/controller/LottoController.java:
```diff
@@ -10,43 +12,30 @@ public class LottoController {
     private final InputView inputView;
     private final OutputView outputView;
+    private final RandomNumberGenerator randomNumberGenerator;
 
```
RandomNumberGenerator은 도메인 모델로 보이는데 도메인 모델이 controller 필드에 존재해도 괜찮을까요?

gabean13 replied:
네 도메인 로직을 담당하는 RandomNumberGenerator가 데이터를 전달하는 Controller에 필드로 있는 건 안된다고 생각해서 수정하였습니다!

hyunssooo commented on src/main/java/model/Lotto.java:
```diff
@@ -1,59 +1,40 @@
 
-    private final List<Integer> lotto;
+    private final Set<Integer> lotto;
 
```
1 ~ 45 숫자 하나를 의미하는 객체를 만들어 보는 것은 어떨까요?

gabean13 replied:
네 그래서 1~45까지의 integer만 저장할 수 있는 LottoNumber라는 객체를 만들고, 해당 객체에서 범위를 체크하는 로직을 만들었습니다. 그런 뒤 Lotto, WinningLotto의 numbers, bonus 전부 LottoNumber로 변경하였습니다. 그래서 범위를 체크하는 중복되는 로직이 삭제될 수 있었습니다

hyunssooo commented on src/main/java/model/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Random;
+import java.util.Set;
+
+public class LottoNumberGenerator implements RandomNumberGenerator {
```
Random은 테스트하기 어렵기 때문에 추상화를 통해 테스트 가능한 구조로 만들기도 하는데요. 
이건 인터페이스가 RandomNumberGenerator 이네요. 
`NumberGenerator`로 추상화를 하고 `RandomNumberGenerator` 를 구현체로 만들어 주는 것은 어떨까요? 
실제로 구현도 그렇게 되어 있는 것 같아요.

hyunssooo commented on src/main/java/model/LottoResult.java:
```diff
@@ -1,31 +1,40 @@
+        }
+
+        return purchaseCount * 1000;
     }
```
depth를 1로 줄여보는 것은 어떨까요? `1000`은 뭘 의미할까요?

gabean13 replied:
넵 Stream으로 해당 로직 수정하였습니다. 1000은 로또 가격을 의미하여 상수로 LOTTO_PRICE로 저장하였습니다.

hyunssooo commented on src/main/java/model/UserLotto.java:
```diff
@@ -2,46 +2,62 @@
+    }
+
+    private int getDuplicateNumber(Lotto lotto, WinningLotto winningLotto) {
+        return winningLotto.calculateDuplicateNumber(lotto.toDto().lotto());
```
> Dto는 도메인의 로직이 필요없이 데이터를 전달할때만 사용하려고 했습니다. (view).

여기는 도메인 모델의 도메인 로직으로 보이는데 dto가 사용되고 있네요?

gabean13 replied:
네..ㅠㅠ 많이 잘못 짠 부분인 것 같습니다.. 😢  해당 부분은 Lotto에서 WinningLotto의 numbers(Lotto객체)를 받아서 같은 필드에 접근할 수 있도록 하는 방법을 통해 구하도록 수정하였습니다.

hyunssooo commented on src/main/java/model/UserLotto.java:
```diff
@@ -2,46 +2,62 @@
-            lottosDto.add(LottoDto.getSortedLottoDto(lotto));
+    private Rank calculateRank(Lotto lotto, WinningLotto winningLotto) {
+        int duplicateNumber = getDuplicateNumber(lotto, winningLotto);
+        if (duplicateNumber == BONUS_REQUIRED_RANK_NUMBER && winningLotto.isBonusMatch(lotto)) {
```
여긴 결국 winningLotto에 lotto를 넘겨서 몇개의 숫자가 맞는지 보너스 숫자가 맞는지를 판단하고 있는데요. 메서드를 두번 호출하고 있네요. 한번으로 Rank 객체를 넘길 수 있지 않을까요?

gabean13 replied:
넵 이 부분은 duplicateNumber는 로또에서 구해오는 것으로 변경하여 해결했습니다

hyunssooo commented on src/main/java/view/OutputView.java:
```diff
@@ -25,11 +28,17 @@ public void printError(String message) {
-                .forEach(System.out::println);
+        for (LottoDto lottoDto : lottosDto) {
+            List<String> sortedList = lottoDto.lotto().stream().sorted().map(String::valueOf).toList();
+            printLotto(sortedList);
```
Stream은 연산마다 개행을 주면 가독성을 높일 수 있습니다. :)

gabean13 replied:
넵 감사합니다 그렇게 수정했습니다!

hyunssooo commented on src/test/java/model/LottoResultTest.java:
```diff
@@ -0,0 +1,50 @@
+    @BeforeEach
+    void setUp() {
+        winningLotto = new WinningLotto(new ArrayList<>(List.of(1, 2, 3, 4, 5, 6)), 7);
+        userLotto = new UserLotto(new FixedNumberGenerator(), 1000);
```
```suggestion
    void setUp() {
        winningLotto = new WinningLotto(List.of(1, 2, 3, 4, 5, 6), 7);
        userLotto = new UserLotto(new FixedNumberGenerator(), 1000);
```

gabean13 replied:
이렇게 수정하였습니다

hyunssooo commented on src/test/java/model/LottoTest.java:
```diff
@@ -1,55 +1,40 @@
-            assertThat(number).isBetween(1, 45);
-        }
-    }
+    void lottoGenerateTest() {
+        Set<Integer> numbers = new HashSet<>(Set.of(1, 2, 3, 4, 5, 45));
```
```suggestion
    void lottoGenerateTest() {
        Set<Integer> numbers = Set.of(1, 2, 3, 4, 5, 45);
```

gabean13 replied:
이렇게 수정하였습니다!

hyunssooo commented on src/test/java/model/UserLottoTest.java:
```diff
@@ -1,28 +1,114 @@
 
 class UserLottoTest {
+    EnumMap<Rank, Integer> ranks;
+    UserLotto userLotto;
```
```suggestion
    Map<Rank, Integer> ranks;
    UserLotto userLotto;
```

gabean13 replied:
이렇게 수정하였습니다! 앞으로는 Interface로 선언하는 습관 들이도록 하겠습니다!

hyunssooo commented on src/test/java/model/WinningLottoTest.java:
```diff
@@ -1,64 +1,56 @@
     @Test
-    void 보너스_볼이_1_이상_45_이하의_정수인지_검사한다() {
-        WinningLotto winningLotto = new WinningLotto("1, 2, 3, 4, 15, 45");
-        org.assertj.core.api.Assertions.assertThatCode(() -> {winningLotto.setBonus("6");})
-                .doesNotThrowAnyException();
+    void isBonusMatchTest() {
+        WinningLotto winningLotto = new WinningLotto(new ArrayList<>(List.of(2, 3, 4, 15, 44, 45)), 1);
+        Lotto trueLotto = new Lotto(new HashSet<>(Set.of(1, 2, 3, 4, 5, 6)));
+        Lotto falseLotto = new Lotto(new HashSet<>(Set.of(10, 2, 3, 4, 5, 6)));
+        org.junit.jupiter.api.Assertions.assertTrue(winningLotto.isBonusMatch(trueLotto));
+        org.junit.jupiter.api.Assertions.assertFalse(winningLotto.isBonusMatch(falseLotto));
```
이런 케이스를 나눠 테스트하면 좋습니다. 가지는 경우, 가지지 않는 경우

gabean13 replied:
네 true인 경우와 false인 경우로 나누어서 테스트 수정하였습니다!!

hyunssooo reviewed (changes requested):
안녕하세요. 가콩!
몇 가지 코멘트 남겼습니다!
함께 고민해 보면 좋을 것 같아요!

gabean13 committed:
refactor : controller에서  numberGenerator 필드 제거

- RandomNumberGenerator -> NumberGenerator로 이름 변경
- LottoNumberGenerator -> RandomNumberGenerator로 이름 변경

gabean13 committed:
refactor : Integer 대신 1 ~ 45 숫자만 저장 가능한 LottoNumber 객체 구현

gabean13 committed:
test : 테스트 분리 및 불필요한 코드 삭제

gabean13 committed:
refactor : for문 stream으로 수정, EnumMap Map으로 변경

gabean13 commented:
말씀 주신 부분 최대한 고민해보고, 이유를 생각해본 뒤에 수정해보려고 노력했습니다.
좋은 리뷰 주셔서 감사합니다 지노! 😄

hyunssooo commented on src/main/java/model/LottoNumber.java:
```diff
@@ -0,0 +1,38 @@
+package model;
+
+import java.util.Objects;
+
+public class LottoNumber {
+    private static final String LOTTO_NUMBER_FORM_ERROR_MESSAGE = "중복되지 않는 1 이상 45 이하의 정수여야합니다.\n";
+    private static final int LOTTO_NAX_SIZE = 45;
+    private static final int LOTTO_MIN_SIZE = 1;
+    private final int number;
+
+    public LottoNumber(int number) {
+        if (number < LOTTO_MIN_SIZE || number > LOTTO_NAX_SIZE) {
+            throw new IllegalArgumentException(LOTTO_NUMBER_FORM_ERROR_MESSAGE);
+        }
+        this.number = number;
+    }
```
해당 클래스는 생성이 굉장히 많은 클래스인 것 같아요. 미리 1 ~ 45까지 생성하고 재사용하는 방법은 없을까요?

hyunssooo replied:
record 로도 표현이 가능하네요! record에 대해서 학습해 보는 것은 어떨까요?

hyunssooo replied:
테스트가 없군요.. ㅠㅠ

hyunssooo commented on src/main/java/model/NumberGenerator.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.Set;
+
+public interface NumberGenerator {
+    Set<Integer> generateNumbers();
+}
```
NumberGenerator 이니 `Set<LottoNumber>`를 리턴해도 괜찮지 않을까요? 더 나아가 Lotto 자체를 리턴해도 괜찮을 것 같아요.

hyunssooo commented on src/main/java/model/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,24 @@
+    private static final int LOTTO_SIZE = 6;
+    private static final int LOTTO_NAX_SIZE = 45;
+    private static final int LOTTO_MIN_SIZE = 1;
+
+    private final Random random = new Random();
+
```
랜덤 클래스는 한번만 생성하고 재사용 가능할 것 같아요.

hyunssooo reviewed (approved):
안녕하세요! 가콩! 
몇 가지 가벼운 코멘트 남겼는데 시간 나실 때 확인 부탁드릴게요!
이번 미션 진행하신다고 고생 많으셨습니다! 이만 merge 하겠습니다!
